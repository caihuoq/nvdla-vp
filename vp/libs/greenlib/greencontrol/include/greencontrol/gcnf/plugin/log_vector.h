//   GreenControl framework
//
// LICENSETEXT
//
//   Copyright (C) 2009 : GreenSocs Ltd
// 	 http://www.greensocs.com/ , email: info@greensocs.com
//
//   Developed by :
//   
//   Michael Ruetz <m.ruetz@tu-bs.de>,
//   Christian Schroeder <schroeder@eis.cs.tu-bs.de>
//     Technical University of Braunschweig, Dept. E.I.S.
//     http://www.eis.cs.tu-bs.de
//
//
// The contents of this file are subject to the licensing terms specified
// in the file LICENSE. Please consult this file for restrictions and
// limitations that may apply.
// 
// ENDLICENSETEXT

// doxygen comments

#ifndef __LOG_VECTOR_H__
#define __LOG_VECTOR_H__

#include "greencontrol/core/log_if.h"

namespace gs {
namespace cnf {


/// This is an extended std::vector with additional log_if methods.
/**
 * This class extends the std::vector by adding the methods demanded by the log_if interface.
 */
template <typename T>
class log_vector
  : public std::vector<T>
  , public log_if
{
public:

  /// Copy-operator.
  log_vector<T>& operator = (const vector<T>& v)
  {
    vector<T>::operator=(v);
    return *this;
  }

  /// Returns a string representation of the vector.
  std::string toString()
  {
    std::stringstream ss;
    ss << getTypeLog() << " = { " << getValueLog() << " }";
    return ss.str();
  }

  /// Returns a string representation of the payload.
  std::string getValueLog()
  {
    std::stringstream ss;
    unsigned int i = 0;

    for(typename vector<T>::iterator it = this->begin(); it != this->end(); ++it)
    {
      i++;
      ss << *it;
      if(i < this->size())
        ss << ", ";
    }

    return ss.str();
  }

  /// Should return the datatype of the vector.
  std::string getTypeLog()
  {
    return std::string("vector<T>");
  }

  /// Should return the name of the vector.
  std::string getNameLog()
  {
    return std::string("N/A");
  }
};


} // end namespace cnf
} // end namespace gs

#endif
