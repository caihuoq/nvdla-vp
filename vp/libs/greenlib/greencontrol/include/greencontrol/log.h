//   GreenControl framework
//
// LICENSETEXT
//
//   Copyright (C) 2009 : GreenSocs Ltd
// 	 http://www.greensocs.com/ , email: info@greensocs.com
//
//   Developed by :
//   
//   Michael Ruetz <m.ruetz@tu-bs.de>,
//   Christian Schroeder <schroeder@eis.cs.tu-bs.de>
//     Technical University of Braunschweig, Dept. E.I.S.
//     http://www.eis.cs.tu-bs.de
//
//
//   This program is free software.
// 
//   If you have no applicable agreement with GreenSocs Ltd, this software
//   is licensed to you, and you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation; either version 2 of the License, or
//   (at your option) any later version.
// 
//   If you have a applicable agreement with GreenSocs Ltd, the terms of that
//   agreement prevail.
// 
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
// 
//   You should have received a copy of the GNU General Public License
//   along with this program; if not, write to the Free Software
//   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
//   02110-1301  USA 
// 
// ENDLICENSETEXT

// doxygen comments

#ifndef __LOG_H__
#define __LOG_H__

#include "greencontrol/gclog/gc_logger.h"
#include "greencontrol/gclog/filter.h"
#include "greencontrol/gclog/output.h"

#endif
