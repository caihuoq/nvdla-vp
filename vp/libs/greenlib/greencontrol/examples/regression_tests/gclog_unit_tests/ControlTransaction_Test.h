//   GreenControl framework
//
// LICENSETEXT
//
//   Copyright (C) 2009 : GreenSocs Ltd
// 	 http://www.greensocs.com/ , email: info@greensocs.com
//
//   Developed by :
//   
//   Michael Ruetz <m.ruetz@tu-bs.de>,
//   Christian Schroeder <schroeder@eis.cs.tu-bs.de>
//     Technical University of Braunschweig, Dept. E.I.S.
//     http://www.eis.cs.tu-bs.de
//
//
// The contents of this file are subject to the licensing terms specified
// in the file LICENSE. Please consult this file for restrictions and
// limitations that may apply.
// 
// ENDLICENSETEXT

// doxygen comments

#ifndef __CONTROLTRANSACTION_TEST_H__
#define __CONTROLTRANSACTION_TEST_H__

#include "greencontrol/log.h"
#include "TestCase.h"
#include "greencontrol/gcnf/plugin/string_vector.h"
#include "greencontrol/gcnf/plugin/log_vector.h"
#include "Command_if_dummy.h"

// GCLOG_RETURN should be enabled
class ControlTransaction_Test : public TestCase
{
public:
  ControlTransaction_Test() {}
  virtual ~ControlTransaction_Test(void) {}

  std::string identify()
  {
    return "ControlTransaction_Test";
  }

  void setup()
  {
    pLog = new GC_Logger();
    pTextOut = new DebugTextfile("ControlTransaction_Test.txt", false);

    pLog->setFilter(pTextOut);
  }

  void destroy()
  {
    delete pLog;
    delete pTextOut;
  }

  // This is for testing the ControlTransaction, LogConfig, log_if, command_if, string_vector and log_vector classes
  void run()
  {
    ControlTransaction c1, c2;
    ControlTransactionHandle ct, ct2;
    LogConfig cfg, cfg2, cfg3;
    std::stringstream ss;

    // Test 1: The LogConfig class gets tested.
    pTextOut->writePure("Test 1\nThe LogConfig class should be printed three times.");
    pTextOut->writePure("The first is set to true, the second to false and the third one is mixed.\n");
    pTextOut->writePure(cfg.toString());
    cfg.reset(false);
    pTextOut->writePure(cfg.toString());
    cfg.setAnyPointer(true);
    cfg.setCmd(true);
    cfg.setLogPointer(true);
    pTextOut->writePure(cfg.toString());

    // Test 2: The different to*String() methods of the CT class are being tested. The LogConfig from the previous test gets activated.
    pTextOut->writePure("\nTest 2\nThe four different toString() methods of the CT are being tested. (toString, toDetailedString, toConfigString, toDebugString)");
    pTextOut->writePure("The LogConfig from the previous test is activated. The output should change accordingly.\n");
    c1.activateConfig(&cfg);
    pTextOut->writePure(c1.toString());
    pTextOut->writePure(c1.toDetailedString());
    pTextOut->writePure(c1.toConfigString());
    pTextOut->writePure(c1.toDebugString());
    c1.deactivateConfig();

    // Test 3: CT hasChanged flag, copy operator, LogConfig/stack
    pTextOut->writePure("\nTest 3\nThe automatic setting of the hasChanged flag is being tested.\n");
    // The hasChanged flag is being tested
    c1.set_mMetaData("Transaction 1");
    c2.set_mMetaData("Transaction 2");
    c1.resetChangeFlag();
    pTextOut->writePure(c1.toDebugString());
    pTextOut->writePure(c2.toDebugString());
    // The copy operator of the CT gets tested.
    pTextOut->writePure("\nThe copy operator of the CT gets tested.\n");
    c2 = c1;
    pTextOut->writePure(c1.toDebugString());
    pTextOut->writePure(c2.toDebugString());
    // The behaviour of the copy operator gets tested, while a LogConfig is activated. This LogConfig object must not be copied.
    pTextOut->writePure("\nThe copy operator of the CT gets tested again, this time with an activated LogConfig. The LogConfig must not be copied\n");
    c2.set_mMetaData("Transaction 2");
    c1.activateConfig(&cfg);
    c2 = c1;
    pTextOut->writePure(c1.toDebugString());
    pTextOut->writePure(c2.toDebugString());
    c1.deactivateConfig();

    // Test 4: The behaviour of the integrated stack gets tested
    pTextOut->writePure("\nTest 4\nThe behaviour of the integrated stack is being tested.\n");
    ct = getCT();
    ct2 = getCT();
    ct->set_mMetaData("Transaction 1");
    ct2->set_mMetaData("Transaction 2");
    c1.activateConfig(&cfg);
    // try to activate a NULL pointer
    std::cout << "\nA warning about activating a NULL pointer should follow:\n";
    ct->activateConfig(NULL);
    // try to activate a LogConfig, that is already in use
    std::cout << "\nA warning about activating an already active LogConfig should follow:\n";
    ct->activateConfig(&cfg);
    c1.deactivateConfig();
    // the normal behaviour of the stack gets tested
    pTextOut->writePure(ct->toDebugString());
    ct->activateConfig(&cfg);
    ct->activateConfig(&cfg2);
    pTextOut->writePure(ct->toDebugString());
    ct->deactivateConfig();
    pTextOut->writePure(ct->toDebugString());
    ct->deactivateConfig();
    // try to deactivate the default LogConfig of the CT
    std::cout << "\nA warning about deactivating the default LogConfig should follow:\n";
    ct->deactivateConfig();
    pTextOut->writePure(ct->toDebugString());

    // Test 5: The copy operator of the LogConfig class gets tested.
    pTextOut->writePure("\nTest 5\nThe copy operator of the LogConfig class is being tested.\n");
    // Two CTs/stacks are getting prepared
    ct = getCT();
    ct2 = getCT();
    ct->set_mMetaData("Transaction 1");
    ct2->set_mMetaData("Transaction 2");
    ct2->resetChangeFlag();
    cfg.reset(false);
    cfg.setSpecifier(true);
    ct->activateConfig(&cfg);
    cfg2.reset(true);
    cfg3.setSpecifier(false);
    ct2->activateConfig(&cfg3);
    ct2->activateConfig(&cfg2);
    pTextOut->writePure(ct->toDebugString());
    pTextOut->writePure(ct2->toDebugString());
    // Copy the LogConfig, the single parts are being copied. The stack must not change.
    pTextOut->writePure("\nThe LogConfigs are being copied, the stack must not change.\n");
    cfg = cfg2;
    pTextOut->writePure(ct->toDebugString());
    pTextOut->writePure(ct2->toDebugString());
    // The default LogConfig gets used.
    pTextOut->writePure("\nThe default LogConfig of Transaction 1 gets modified.\n");
    ct->deactivateConfig();
    ct->get_mLogConfig()->setSpecifier(false);
    pTextOut->writePure(ct->toDebugString());
    *(ct->get_mLogConfig()) = cfg2;
    pTextOut->writePure(ct->toDebugString());
    ct2->deactivateConfig();
    ct2->deactivateConfig();
    
    // Test 6: The command_if interface gets tested.
    pTextOut->writePure("\nTest 6\nThe command_if interface gets tested.");
    pTextOut->writePure("The first time no command_if object is set, which results in warnings. The second time the command_if is set.\n");
    Command_if_dummy cmdIf;
    ct = getCT();
    // There is no command_if set, so a warning should be printed.
    pTextOut->write(ct->getSenderName());
    pTextOut->write(ct->getCommandName());
    pTextOut->write(ct->getCommandDescription());
    pTextOut->writePure(ct->toDebugString());
    // now a command_if gets set
    ct->set_mCmdIf(&cmdIf);
    pTextOut->write(ct->getSenderName());
    pTextOut->write(ct->getCommandName());
    pTextOut->write(ct->getCommandDescription());
    pTextOut->writePure(ct->toDebugString());

    // Test 7: The log_if interface gets tested. Therefore the string_vector and log_vector classes are being used.
    pTextOut->writePure("\nTest 7\nThe log_if interface gets tested. Therefore the string_vector and log_vector classes are being used.\n");
    gs::cnf::string_vector stringVec;
    std::vector<std::string> vec;
    vec.push_back("Added to vec");
    // the copy operator gets tested
    stringVec = vec;
    pTextOut->write(stringVec.toString());
    stringVec.push_back("Added to stringVec");
    ct->set_mLogPointer(&stringVec);
    pTextOut->writePure(ct->toDebugString());

    // the log_vector is being used
    gs::cnf::log_vector<int> logVec;
    std::vector<int> intVec;
    intVec.push_back(42);
    // test the copy operator
    logVec = intVec;
    pTextOut->write(logVec.toString());
    logVec.push_back(-5);
    logVec.push_back(85);
    ct->set_mLogPointer(&logVec);
    pTextOut->writePure(ct->toDebugString());

    // test/demonstrate the log_vector with different datatypes
    gs::cnf::log_vector<void*> addrVec;
    addrVec.push_back(this);
    addrVec.push_back(&logVec);
    addrVec.push_back(pTextOut);
    pTextOut->write(addrVec.toString());

    gs::cnf::log_vector<std::string> logStrVec;
    logStrVec.push_back("Test String");
    logStrVec.push_back(stringVec[0]);
    logStrVec.push_back(stringVec.toString());
    pTextOut->write(logStrVec.toString());
  }

protected:
  GC_Logger* pLog;
  DebugTextfile* pTextOut;
};

#endif
