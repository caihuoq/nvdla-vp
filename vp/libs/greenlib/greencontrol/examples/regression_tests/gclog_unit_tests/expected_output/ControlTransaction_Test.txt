Tuesday, 16 February 2010 - 13:06:06
Debug text output used with GC_Logger.
Starting log.

Test 1
The LogConfig class should be printed three times.
The first is set to true, the second to false and the third one is mixed.

    LogConfig 0xbfe8d834:
      m_bHasChanged      = 0
      m_bStandardSetting = 1
      m_bService         = 1
      m_bTarget          = 1
      m_bCmd             = 1
      m_bAnyPointer      = 1
      m_bAnyPointer2     = 1
      m_bAnyUint         = 1
      m_bSpecifier       = 1
      m_bValue           = 1
      m_bID              = 1
      m_bError           = 1
      m_bMetaData        = 1
      m_bLogPointer      = 1
      m_bCmdIf           = 1
      m_bLogConfig       = 1
      m_pNext            = 0
    LogConfig 0xbfe8d834:
      m_bHasChanged      = 0
      m_bStandardSetting = 0
      m_bService         = 0
      m_bTarget          = 0
      m_bCmd             = 0
      m_bAnyPointer      = 0
      m_bAnyPointer2     = 0
      m_bAnyUint         = 0
      m_bSpecifier       = 0
      m_bValue           = 0
      m_bID              = 0
      m_bError           = 0
      m_bMetaData        = 0
      m_bLogPointer      = 0
      m_bCmdIf           = 0
      m_bLogConfig       = 0
      m_pNext            = 0
    LogConfig 0xbfe8d834:
      m_bHasChanged      = 1
      m_bStandardSetting = 0
      m_bService         = 0
      m_bTarget          = 0
      m_bCmd             = 1
      m_bAnyPointer      = 1
      m_bAnyPointer2     = 0
      m_bAnyUint         = 0
      m_bSpecifier       = 0
      m_bValue           = 0
      m_bID              = 0
      m_bError           = 0
      m_bMetaData        = 0
      m_bLogPointer      = 1
      m_bCmdIf           = 0
      m_bLogConfig       = 0
      m_pNext            = 0

Test 2
The four different toString() methods of the CT are being tested. (toString, toDetailedString, toConfigString, toDebugString)
The LogConfig from the previous test is activated. The output should change accordingly.

  Transaction 0xbfe8d7c0:
    mService     = 0 (NO_SERVICE)
    mTarget      = 0
    mCmd         = 0
    mAnyPointer  = 0
    mAnyPointer2 = 0
    mAnyUint     = 36
    mSpecifier   = 
    mValue       = 
    mID          = 0x31
    mMetaData    = 
    mError       = 0
    mLogPointer  = 0
    mCmdIf       = 0
    mLogConfig   = 0xbfe8d834
  Transaction 0xbfe8d7c0:
    mService     = 0 (NO_SERVICE)
    mTarget      = 0
    mCmd         = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  = 0
    mAnyPointer2 = 0
    mAnyUint     = 36
    mSpecifier   = 
    mValue       = 
    mID          = 0x31 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    = 
    mError       = 0
    mLogPointer  = 0
    mCmdIf       = 0
    mLogConfig   = 0xbfe8d834
  Transaction 0xbfe8d7c0:
    mCmd         = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  = 0
    mLogPointer  = 0

  Transaction 0xbfe8d7c0:
    mHasChanged      = 0
    mService     (0) = 0 (NO_SERVICE)
    mTarget      (0) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (0) = 0
    mAnyUint     (0) = 36
    mSpecifier   (0) = 
    mValue       (0) = 
    mID          (0) = 0x31 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (0) = 
    mError       (0) = 0
    mLogPointer  (1) = 0
    mCmdIf       (0) = 0
    mLogConfig   (0) = 0xbfe8d834 (hasChanged: 1)
    -> pNext     (1) = 0x93a8470 -> 0

Test 3
The automatic setting of the hasChanged flag is being tested.

  Transaction 0xbfe8d7c0:
    mHasChanged      = 0
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 36
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x31 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a8470 (hasChanged: 0)
    -> pNext     (0) = 0
  Transaction 0xbfe8d784 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 3740576
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x38fff4 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 2
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a2740 (hasChanged: 1)
    -> pNext     (0) = 0

The copy operator of the CT gets tested.

  Transaction 0xbfe8d7c0:
    mHasChanged      = 0
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 36
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x31 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a8470 (hasChanged: 0)
    -> pNext     (0) = 0
  Transaction 0xbfe8d784:
    mHasChanged      = 0
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 36
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x31 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a2740 (hasChanged: 1)
    -> pNext     (0) = 0

The copy operator of the CT gets tested again, this time with an activated LogConfig. The LogConfig must not be copied

  Transaction 0xbfe8d7c0:
    mHasChanged      = 0
    mService     (0) = 0 (NO_SERVICE)
    mTarget      (0) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (0) = 0
    mAnyUint     (0) = 36
    mSpecifier   (0) = 
    mValue       (0) = 
    mID          (0) = 0x31 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (0) = Transaction 1
    mError       (0) = 0
    mLogPointer  (1) = 0
    mCmdIf       (0) = 0
    mLogConfig   (0) = 0xbfe8d834 (hasChanged: 1)
    -> pNext     (1) = 0x93a8470 -> 0
  Transaction 0xbfe8d784:
    mHasChanged      = 0
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 36
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x31 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a2740 (hasChanged: 1)
    -> pNext     (0) = 0

Test 4
The behaviour of the integrated stack is being tested.

  Transaction 0x93a2760 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a27a0 (hasChanged: 1)
    -> pNext     (0) = 0
  Transaction 0x93a2760 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0xbfe8d818 (hasChanged: 0)
    -> pNext     (1) = 0xbfe8d834 -> 0x93a27a0 -> 0
  Transaction 0x93a2760 returning:
    mHasChanged      = 1
    mService     (0) = 0 (NO_SERVICE)
    mTarget      (0) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (0) = 0
    mAnyUint     (0) = 543516756
    mSpecifier   (0) = 
    mValue       (0) = 
    mID          (0) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (0) = Transaction 1
    mError       (0) = 0
    mLogPointer  (1) = 0
    mCmdIf       (0) = 0
    mLogConfig   (0) = 0xbfe8d834 (hasChanged: 1)
    -> pNext     (1) = 0x93a27a0 -> 0
  Transaction 0x93a2760 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a27a0 (hasChanged: 1)
    -> pNext     (0) = 0

Test 5
The copy operator of the LogConfig class is being tested.

  Transaction 0x93a65e8 returning:
    mHasChanged      = 1
    mService     (0) = 0 (NO_SERVICE)
    mTarget      (0) = 0
    mCmd         (0) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (0) = 0
    mAnyPointer2 (0) = 0
    mAnyUint     (0) = 543516756
    mSpecifier   (1) = 
    mValue       (0) = 
    mID          (0) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (0) = Transaction 1
    mError       (0) = 0
    mLogPointer  (0) = 0
    mCmdIf       (0) = 0
    mLogConfig   (0) = 0xbfe8d834 (hasChanged: 1)
    -> pNext     (1) = 0x93a2720 -> 0
  Transaction 0x93a2760:
    mHasChanged      = 0
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 2
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0xbfe8d818 (hasChanged: 0)
    -> pNext     (1) = 0xbfe8d7fc -> 0x93a6490 -> 0

The LogConfigs are being copied, the stack must not change.

  Transaction 0x93a65e8 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0xbfe8d834 (hasChanged: 0)
    -> pNext     (1) = 0x93a2720 -> 0
  Transaction 0x93a2760:
    mHasChanged      = 0
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 2
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0xbfe8d818 (hasChanged: 0)
    -> pNext     (1) = 0xbfe8d7fc -> 0x93a6490 -> 0

The default LogConfig of Transaction 1 gets modified.

  Transaction 0x93a65e8 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (0) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a2720 (hasChanged: 1)
    -> pNext     (0) = 0
  Transaction 0x93a65e8 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = Transaction 1
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a2720 (hasChanged: 0)
    -> pNext     (0) = 0

Test 6
The command_if interface gets tested.
The first time no command_if object is set, which results in warnings. The second time the command_if is set.

User output: Warning: API/Plugin doesn't implement command_if!
User output: Warning: API/Plugin doesn't implement command_if!
User output: Warning: API/Plugin doesn't implement command_if!
  Transaction 0x93a6430 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (Warning: API/Plugin doesn't implement command_if!)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Warning: API/Plugin doesn't implement command_if!)
    mMetaData    (1) = 
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0
    mLogConfig   (1) = 0x93a27a0 (hasChanged: 1)
    -> pNext     (0) = 0
User output: Command_if_dummy instance name
User output: CMD_0
User output: Doesn't do anything.
  Transaction 0x93a6430 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (CMD_0)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Command_if_dummy instance name)
    mMetaData    (1) = 
    mError       (1) = 0
    mLogPointer  (1) = 0
    mCmdIf       (1) = 0xbfe8d96c
    mLogConfig   (1) = 0x93a27a0 (hasChanged: 1)
    -> pNext     (0) = 0

Test 7
The log_if interface gets tested. Therefore the string_vector and log_vector classes are being used.

User output: vector<string> = { Added to vec }
  Transaction 0x93a6430 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (CMD_0)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Command_if_dummy instance name)
    mMetaData    (1) = 
    mError       (1) = 0
    mLogPointer  (1) = 0xbfe8d880 ( vector<string> = { Added to vec, Added to stringVec } )
    mCmdIf       (1) = 0xbfe8d96c
    mLogConfig   (1) = 0x93a27a0 (hasChanged: 1)
    -> pNext     (0) = 0
User output: vector<T> = { 42 }
  Transaction 0x93a6430 returning:
    mHasChanged      = 1
    mService     (1) = 0 (NO_SERVICE)
    mTarget      (1) = 0
    mCmd         (1) = 0 (CMD_0)
    mAnyPointer  (1) = 0
    mAnyPointer2 (1) = 0
    mAnyUint     (1) = 543516756
    mSpecifier   (1) = 
    mValue       (1) = 
    mID          (1) = 0x93a1688 (Command_if_dummy instance name)
    mMetaData    (1) = 
    mError       (1) = 0
    mLogPointer  (1) = 0xbfe8d870 ( vector<T> = { 42, -5, 85 } )
    mCmdIf       (1) = 0xbfe8d96c
    mLogConfig   (1) = 0x93a27a0 (hasChanged: 1)
    -> pNext     (0) = 0
User output: vector<T> = { 0x93a1688, 0xbfe8d870, 0x93a61a0 }
User output: vector<T> = { Test String, Added to vec, vector<string> = { Added to vec, Added to stringVec } }

Regular end of log.
Finished on Tuesday, 16 February 2010 - 13:06:06
Closing log.